### **Problem** - Which industry is predicted to have the highest growth of releases in 5 years? which will have the largest decline.

# Exploratory data Analysis

# **Data explanation**
### **What NPRI is ?** **Dataset understanding** ?
NPRI is the National Pollutant Release Inventory in Canada which tracks over 300 pollutants from over 7,000 facilities across Canada.

We have dataset provided by NPRI (National Pollutant Release Inventory) which shows the amount of realeses by different industries from the year 2014 to 2022. moreover, the name of the companies and the type of industries corresponding to the type of realeses they are producing as well as the Province where each industry and company is located.

Loading the NPRI dataset using pandas(used for dataframes) library.

In [957]:
import pandas as pd
df1 = pd.read_csv("Npri1.csv")


<ipython-input-957-50574718b5ed>:2: DtypeWarning:

Columns (16) have mixed types. Specify dtype option on import or set low_memory=False.



Checking the structure of the dataset

*   Shape( ) - used to check the number of rows and columns in the dataset.

In [958]:
df1.shape

(537550, 18)

We have 537550 rows and 18 columns

# Dataset Features


1.   **NpriID**: NPRI Identification Number.
2.   **Company Name**: Name of the company.
3.   **FacilityName**: Facility Name.
4.   **NPRI_Report_ReportYear**: Reporting year.
5.   **Quantity**: Quantity.
6.   **Cas**: Chemical Abstracts Service number.
7.   **Detail_NPRI_Substanceinfo_NameEn**: Substance name.
8.   **Group**: Describes the types of substance report. Higher level than     category.
9.   **Category**: Describes the specific category of the substance report. These are sub-groups.
10.   **Latitude**: Latitude.
11.  **Longitude**: Longitude.
12.  **ProvinceEn**: Province name.
13.  **Detail_UnitType_NameEn**: Units.
14.  **NAICSPrimary**: North American Industry Classification System.
15.  **NAICSTitleEn**: NAICS english name.
16.  **Substance comments by category**: substance comments by category.
17.  **Substance comments for disposals and all media**: Substance comments for disposals and all all media. These comments are by group rather than category.
18.  **Number of employees**: Number of employees in the company.










**Making a copy of the Original dataset**
* To avoid doing changes to the original dataset.

In [959]:
data = df1.copy()
data

NpriID                        Company name  \
0        28940          Les Planchers Mercier inc.   
1        28940          Les Planchers Mercier inc.   
2        28940          Les Planchers Mercier inc.   
3        28940          Les Planchers Mercier inc.   
4        28940          Les Planchers Mercier inc.   
...        ...                                 ...   
537545   33064      Deltastream Energy Corporation   
537546   33068      Deltastream Energy Corporation   
537547   33075      Deltastream Energy Corporation   
537548  500352  Canadian Natural Resources Limited   
537549  500352  Canadian Natural Resources Limited   

                                    FacilityName  NPRI_Report_ReportYear  \
0                     Les Planchers Mercier inc.                    2021   
1                     Les Planchers Mercier inc.                    2020   
2                     Les Planchers Mercier inc.                    2021   
3                     Les Planchers Mercier inc.                    2020   
4                     Les Planchers Mercier inc.                    2021   
...                                          ...                     ...   
537545            Delta Nipisi 03-29-76-05w5 MWB                    2021   
537546              Delta Canal 3-29-70-23w4 MWB                    2021   
537547         Delta Canal 14-16-70-23w4 Pro MWB                    2021   
537548  Wapiti Compressor Station 07-24-068-07W6                    2018   
537549  Wapiti Compressor Station 07-24-068-07W6                    2020   

        Quantity        Cas             Detail_NPRI_SubstanceInfo_NameEn  \
0          22.63   NA - M08                     Total particulate matter   
1          16.67   NA - M08                     Total particulate matter   
2          15.94   630-08-0                              Carbon monoxide   
3          13.25   630-08-0                              Carbon monoxide   
4          12.48   NA - M09  PM10 - Particulate Matter <= 10 Micrometers   
...          ...        ...                                          ...   
537545      0.00   NA - M16            Volatile Organic Compounds (VOCs)   
537546      0.00   NA - M16            Volatile Organic Compounds (VOCs)   
537547      0.00   NA - M16            Volatile Organic Compounds (VOCs)   
537548      0.00  7446-09-5                              Sulphur dioxide   
537549      0.00  7446-09-5                              Sulphur dioxide   

                  Group            Category   Latitude   Longitude ProvinceEn  \
0       Releases to Air       Stack / Point  45.869119  -72.530779     Quebec   
1       Releases to Air       Stack / Point  45.869119  -72.530779     Quebec   
2       Releases to Air       Stack / Point  45.869119  -72.530779     Quebec   
3       Releases to Air       Stack / Point  45.869119  -72.530779     Quebec   
4       Releases to Air       Stack / Point  45.869119  -72.530779     Quebec   
...                 ...                 ...        ...         ...        ...   
537545  Releases to Air  Storage / Handling  55.608903 -114.744156    Alberta   
537546  Releases to Air  Storage / Handling  55.084040 -113.478938    Alberta   
537547  Releases to Air  Storage / Handling  55.065795 -113.453519    Alberta   
537548  Releases to Air       Stack / Point  54.899664 -118.925345    Alberta   
537549  Releases to Air       Stack / Point  54.899664 -118.925345    Alberta   

       Detail_UnitType_NameEn  NAICSPrimary  \
0                      tonnes        321999   
1                      tonnes        321999   
2                      tonnes        321999   
3                      tonnes        321999   
4                      tonnes        321999   
...                       ...           ...   
537545                 tonnes        211110   
537546                 tonnes        211110   
537547                 tonnes        211110   
537548                 tonnes        211110   
537549                 tonnes        211110

# Dropping and renaming the columns

In [960]:
#Removing unnecessary columns from the dataset which are not helpful in solving the problem we have.
data = data.drop(columns = ["NpriID" ,"Company name", "FacilityName", "Cas", "Category", "NAICSPrimary","Latitude",	"Longitude", "Substance comments by category for 2022", "Substance comments for disposals and all media for 2022"])

In [961]:
#Giving readable names to the features
data =  data.rename(columns = {"NPRI_Report_ReportYear": "Reporting year","Quantity":"Quantity of realese","Detail_NPRI_SubstanceInfo_NameEn": "Substance name", "ProvinceEn":"Province","Detail_UnitType_NameEn": "Unit","NAICSTitleEn":"Industry"})

In [962]:
#Checking for the count of unique value in unit column using value_counts function
data["Unit"].value_counts()

tonnes    448698
kg         84343
g TEQ       2508
grams       2001
Name: Unit, dtype: int64

# Handling the missing values

Here we are checking the total null values in each column

In [963]:
data.isnull().sum()

Reporting year           0
Quantity of realese    954
Substance name           0
Group                    0
Province                 0
Unit                     0
Industry                 0
Number of employees    578
dtype: int64

As we can see that we have missing values in number of employees and quantity column of the dataset

Missing Values, how to handle

Option 1: Fill NaN with Mean Value
Filling NaNs with the mean value is also not sufficient and naive, and doesn't seems to be a good option.

Option 2: Fill NaN with Last Value with .ffill()
Filling NaNs with the last value could be bit better.

Option 3: Fill NaN with Linearly Interpolated Value with .interpolate()
Filling NaNs with the interpolated values is the best option in this small examlple but it requires knowledge of the neighouring value



In [964]:
data['Quantity of realese'] = data['Quantity of realese'].interpolate()
data['Number of employees'] = data['Number of employees'].interpolate()

For handling missing values, we are using interpolation.
Interpolation in Python is a technique used to estimate unknown data points between two known data points. In Python, Interpolation is a technique mostly used to impute missing values in the data frame or series while preprocessing data.

In [965]:
#Checking the shape of the dataset
data.isnull().sum()

Reporting year         0
Quantity of realese    0
Substance name         0
Group                  0
Province               0
Unit                   0
Industry               0
Number of employees    0
dtype: int64

Now we do not have missing value in any column

**Checking the issues in the column's unique values**

In [966]:
# checking for the uniques value in the unit column
data["Unit"].unique()

array(['tonnes', 'kg', 'g TEQ', 'grams'], dtype=object)

In [967]:
# checking for the uniques value in the Industry column
data["Industry"].unique()

array(['All other miscellaneous wood product manufacturing',
       'All other basic inorganic chemical manufacturing',
       'Motor vehicle metal stamping', 'Material recovery facilities',
       'Other basic organic chemical manufacturing',
       'Plastic pipe and pipe fitting manufacturing',
       'All other miscellaneous chemical product manufacturing',
       'All other miscellaneous manufacturing',
       'Plastic plumbing fixture manufacturing',
       'Other electric power generation', 'Gold and silver ore mining',
       'Cold-rolled steel shape manufacturing',
       'Paint and coating manufacturing',
       'Commercial bakeries and frozen bakery product manufacturing',
       'Coating, engraving, cold and heat treating and allied activities',
       'Polystyrene foam product manufacturing',
       'Paper bag and coated and treated paper manufacturing',
       'Aluminum rolling, drawing, extruding and alloying',
       'Truck trailer manufacturing',
       'All other non-m

In [968]:
# checking for the uniques value in the Reporting year column
data["Reporting year"].unique()

array([2021, 2020, 2019, 2018, 2022, 2017, 2016, 2015, 2014])

In [969]:
# checking for the uniques value in the Substance name column
data["Substance name"].unique()

array(['Total particulate matter', 'Carbon monoxide',
       'PM10 - Particulate Matter <= 10 Micrometers',
       'Nitrogen oxides (expressed as nitrogen dioxide)',
       'PM2.5 - Particulate Matter <= 2.5 Micrometers',
       'Volatile Organic Compounds (VOCs)', 'Sulphur dioxide',
       'Hexavalent chromium (and its compounds)', 'Hydrochloric acid',
       'Ammonia (total)', 'n-Hexane', 'Manganese (and its compounds)',
       'Formaldehyde', 'Zinc (and its compounds)', 'Acrolein', 'Toluene',
       'Mercury (and its compounds)', 'Speciated VOCs (62 substances)',
       'Sulphuric acid', '2-Butoxyethanol',
       'Aluminum oxide (fibrous forms only)', 'Isopropyl alcohol',
       'Styrene', 'Lead (and its compounds)',
       'Arsenic (and its compounds)', 'Cadmium (and its compounds)',
       'Phosphorus (total)', 'Cobalt (and its compounds)',
       'Nitrate ion in solution at pH >= 6.0',
       'Selenium (and its compounds)', 'Copper (and its compounds)',
       'Cyanides (ionic)',

In [970]:
# checking for the uniques value in the Province column
data["Province"].unique()

array(['Quebec', 'Ontario', 'Saskatchewan', 'British Columbia',
       'Manitoba', 'Alberta', 'New Brunswick', 'Nova Scotia', 'Nunavut',
       'Newfoundland and Labrador', 'Prince Edward Island', 'Yukon',
       'Northwest Territories'], dtype=object)

From the above code we can see that there is no unacceptable values in the column

# Handling "Unit" column and "Quantity of realese" column
### Addition of new column named as "Target_column" and removal of "Quantity of realese" column

According to the problem, we are trying to solve we will have to predict that which industry will realese most of the gases or realeses. So, quantity of realeses needs to be our target column But the problem is that the values in the **Quantity of realese** column are in different units like kg, grams, tonnes and g TEQ. That is why, we will have to convert all that values into Single unit.

Creating a column named as **Target_column** with initial value of **0.0** in each row.
After that, converting all the values corresponds to g TEQ, grams and tonnes in the **Quantity of realese** column into kg (kilograms) and filling them in to Target_column.
In the next step we will drop the Quantity of realese column as we do not need it anymore because we have made a new column**(Target_column)** with values of same unit.

In [971]:
# making a copy of the dataset for visualization
data_visual = data.copy()
# converting tonnes, grams, g TEQ to kg and also changing the corresponding values in the quantity of realese column
data.loc[(data["Unit"] == "kg"), "Quantity of realese"] = data["Quantity of realese"]
data.loc[(data["Unit"] == "tonnes"), "Quantity of realese"] = data["Quantity of realese"] * 1000
data.loc[(data["Unit"] == "grams"), "Quantity of realese"] = data["Quantity of realese"] / 1000
data.loc[(data["Unit"] == "g TEQ"), "Quantity of realese"] = data["Quantity of realese"] / 1000
data["Unit"] = data["Unit"].replace({"tonnes" : "kg","g TEQ" : "kg","grams" : "kg"})
data["Unit"].value_counts()


kg    537550
Name: Unit, dtype: int64

In [972]:
# makaing a copy of the data for making box plot because after converting units in kg
data_box = data.copy()

# **Data visualization**

* px.pie(): This function creates a pie chart using Plotly Express. we specify the DataFrame containing the data (data), the column for the values ('Quantity of realese'), the column for the category labels ('Province'), and providing a title for the pie chart.
* fig.update_traces(): This method updates the properties of the pie chart traces (slices). In this case, it positions the text labels outside the pie slices and specifies the information to be displayed in the labels (percentage and label).
* fig.show(): This line displays the pie chart

In [ ]:
import plotly.express as px
# Create a pie chart of Province percentages
fig = px.pie(data, values='Quantity of realese', names='Province', title='Percentage of Releases by Province')
# Add labels to the pie chart
fig.update_traces(textposition='outside', textinfo='percent+label')
# Show the pie chart
fig.show()


**Making bins of Quantity of realese column**

Making the bins of Target_column in the categories of Low, medium, high, and very high.

In [ ]:
# finding the beggining of the bin
data["Quantity of realese"].min()

In [ ]:
# finding the end of the bin
data["Quantity of realese"].max()

In [ ]:
import pandas as pd
#Creating bins for Target_column
bins = [-337320.0, 0.0, 16086629, 24129943.5,4128453559.0]
labels = ["Low", "Medium", "High", "Very High"]
data["Quantity of realese"] = pd.cut(data["Quantity of realese"], bins=bins, labels=labels) #Using cut function from pandas to make bins
data["Quantity of realese"] = data["Quantity of realese"].replace({"Low" : 0,"Medium" : 1,"High" : 2,"Very High" : 3})
data["Quantity of realese"].value_counts()


# Data Sepration

Separating the dataset in to datasets using reoporting year column. One dataset will have data from 2014 to 2021 year and one dataset will have data of 2022 year only

* **data_2014_2021**: This subset contains rows from data where the "Reporting
year" is less than 2022, effectively covering the years 2014 to 2021, inclusive.
* **data_2022**: This subset contains rows from data where the "Reporting year" is equal to 2022, effectively covering only the year 2022.






In [ ]:
# variable data_2014_2021 have data from 2014 to 2021 year
data_2014_2021 = data[data["Reporting year"] < 2022]
# variable data_2022 have data of 2022 year
data_2022 = data[data["Reporting year"] == 2022]
# Getting confirmation using unique function on date_2014_2022 and date_2022 variable to know that our data is seprated.
data_2014_2021["Reporting year"].unique()
data_2022["Reporting year"].unique()
data_2014_2021["Industry"].value_counts()

# Handling outliers

* import plotly.express as px: Import the Plotly Express library as px.
* px.scatter(): This function creates a scatter plot using Plotly Express.

In [ ]:
import plotly.express as px
# Create a scatter plot using Plotly Express
fig = px.scatter(data_box, x="Quantity of realese", y="Number of employees", color="Reporting year", title="Outliers in Number of Employees by Reporting Year")
# Show the plot
fig.show()

Below we are using plotly for making a box plot to detecting outliers
from the box plot we can see that there are some industries with high quantity of realese in 2022 year which are uncommon as compared to other values in other years. So, i am considering these values as outliers.

In [ ]:
import plotly.express as px
# Create a box plot using Plotly Express
fig = px.box(data_box, x="Reporting year", y="Quantity of realese", title="Quantity of realese by Industry")
# Show the plot
fig.show()

Prediction releases for redemption services in the year 2022.

In [ ]:
# Creating a dictionary to store dataframes for each industry
industry_data = {}
# Getting unique industries
unique_industries = data_2014_2021["Industry"].unique()
# Looping through each industry and creating a dataframe
for industry in unique_industries:
  industry_data[industry] = data_2014_2021[data_2014_2021["Industry"] == industry]
Remediation_services_2014_21 = industry_data["Remediation services"]


industry_data_2022 = {}
# Getting unique industries
unique_industries = data_2022["Industry"].unique()
# Looping through each industry and creating a dataframe
for industry in unique_industries:
  industry_data_2022[industry] = data_2022[data_2022["Industry"] == industry]
Remediation_services_2022 = industry_data_2022["Remediation services"]


# importing label encoder from sklearn. preprocessing to convert datatype of each column ot numerical values.
from sklearn.preprocessing import LabelEncoder
# Extracting columns which have object data type
cat_cols = [col for col in Remediation_services_2022.columns if Remediation_services_2022[col].dtype == "object"  or data[col].dtype == "category"]
#Running a for loop on each column and converting them to numerical using fir.transform function of label encoder.
for col in cat_cols:
   Remediation_services_2022[col] = LabelEncoder().fit_transform(Remediation_services_2022[col])
Remediation_services_2022

# importing label encoder from sklearn. preprocessing to convert datatype of each column ot numerical values.
from sklearn.preprocessing import LabelEncoder
# Extracting columns which have object data type
cat_cols = [col for col in Remediation_services_2014_21.columns if Remediation_services_2014_21[col].dtype == "object"  or data[col].dtype == "category"]
#Running a for loop on each column and converting them to numerical using fir.transform function of label encoder.
for col in cat_cols:
   Remediation_services_2014_21[col] = LabelEncoder().fit_transform(Remediation_services_2014_21[col])


In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
# Splitting the data_Remediation_inverse_14 into training and testing sets
X_train23, X_test23, y_train23, y_test23 = train_test_split(Remediation_services_2014_21.drop('Quantity of realese',axis = 1),Remediation_services_2014_21['Quantity of realese'],test_size=0.25,random_state=42
)
X_train24, X_test24, y_train24, y_test24 = train_test_split(
    Remediation_services_2022.drop("Quantity of realese", axis = 1), Remediation_services_2022["Quantity of realese"], test_size = 0.25, random_state = 42)
# Creating a KNN classifier
knn = KNeighborsClassifier(n_neighbors=3)

# Training the model
knn.fit(X_train23, y_train23)

# Predicting the labels for the test data
y_pred24 = knn.predict(X_test24)


# Predicting which industry will have high releases in 2022
predicted_releases = knn.predict(X_test24)

print("Oil and gas extraction (except oil sands) will have medium realeses in 2022 year based on its past ",predicted_releases)

Predicting releases for oil and gas extraction (except oil sands)

In [ ]:
oil_gas_2014_21 = industry_data["Oil and gas extraction (except oil sands)"]
oil_gas_2022 = industry_data_2022["Oil and gas extraction (except oil sands)"]
# importing label encoder from sklearn. preprocessing to convert datatype of each column ot numerical values.
from sklearn.preprocessing import LabelEncoder
# Extracting columns which have object data type
cat_cols = [col for col in oil_gas_2014_21.columns if oil_gas_2014_21[col].dtype == "object"  or oil_gas_2014_21[col].dtype == "category"]
#Running a for loop on each column and converting them to numerical using fir.transform function of label encoder.
for col in cat_cols:
  oil_gas_2014_21[col] = LabelEncoder().fit_transform(oil_gas_2014_21[col])


# importing label encoder from sklearn. preprocessing to convert datatype of each column ot numerical values.
from sklearn.preprocessing import LabelEncoder
# Extracting columns which have object data type
cat_cols = [col for col in oil_gas_2022.columns if oil_gas_2022[col].dtype == "object"  or oil_gas_2022[col].dtype == "category"]
#Running a for loop on each column and converting them to numerical using fir.transform function of label encoder.
for col in cat_cols:
   oil_gas_2022[col] = LabelEncoder().fit_transform(oil_gas_2022[col])



In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
# Splitting the data_Remediation_inverse_14 into training and testing sets
X_train14, X_test14, y_train14, y_test14 = train_test_split(oil_gas_2014_21.drop('Quantity of realese', axis=1),oil_gas_2014_21['Quantity of realese'],test_size=0.25,random_state=42)
X_train22, X_test22, y_train22, y_test22 = train_test_split(oil_gas_2022.drop('Quantity of realese', axis=1), oil_gas_2022['Quantity of realese'],test_size=0.25,random_state=42)

# Creating a KNN classifier
knn = KNeighborsClassifier(n_neighbors=3)

# Training the model
knn.fit(X_train14, y_train14)

# Predicting the labels for the test data
y_pred22 = knn.predict(X_test22)

# Evaluating the accuracy of the model
accuracy = accuracy_score(y_test22, y_pred22)
print('Accuracy:', accuracy)

# Predicting which industry will have high releases in 2022
predicted_releases = knn.predict(X_test22)
print("Oil and gas extraction (except oil sands) will have medium realeses in 2022 year based on its past ")

Predicting releases for

In [ ]:
Petroleum_refineries_2014_21 = industry_data["Petroleum refineries"]
Petroleum_refineries_2022 = industry_data_2022["Petroleum refineries"]
# importing label encoder from sklearn. preprocessing to convert datatype of each column ot numerical values.
from sklearn.preprocessing import LabelEncoder
# Extracting columns which have object data type
cat_cols = [col for col in Petroleum_refineries_2014_21.columns if Petroleum_refineries_2014_21[col].dtype == "object"  or Petroleum_refineries_2014_21[col].dtype == "category"]
#Running a for loop on each column and converting them to numerical using fir.transform function of label encoder.
for col in cat_cols:
  Petroleum_refineries_2014_21[col] = LabelEncoder().fit_transform(Petroleum_refineries_2014_21[col])


# importing label encoder from sklearn. preprocessing to convert datatype of each column ot numerical values.
from sklearn.preprocessing import LabelEncoder
# Extracting columns which have object data type
cat_cols = [col for col in Petroleum_refineries_2022.columns if Petroleum_refineries_2022[col].dtype == "object"  or Petroleum_refineries_2022[col].dtype == "category"]
#Running a for loop on each column and converting them to numerical using fir.transform function of label encoder.
for col in cat_cols:
   Petroleum_refineries_2022[col] = LabelEncoder().fit_transform(Petroleum_refineries_2022[col])

from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
# Splitting the data_Remediation_inverse_14 into training and testing sets
X_train15, X_test15, y_train15, y_test15 = train_test_split(Petroleum_refineries_2014_21.drop('Quantity of realese', axis=1),Petroleum_refineries_2014_21['Quantity of realese'],test_size=0.25,random_state=42)
X_train23, X_test23, y_train23, y_test23 = train_test_split(Petroleum_refineries_2022.drop('Quantity of realese', axis=1), Petroleum_refineries_2022['Quantity of realese'],test_size=0.25,random_state=42)

# Creating a KNN classifier
knn = KNeighborsClassifier(n_neighbors=3)

# Training the model
knn.fit(X_train15, y_train15)

# Predicting the labels for the test data
y_pred23 = knn.predict(X_test23)

# Evaluating the accuracy of the model
accuracy = accuracy_score(y_test23, y_pred23)
print('Accuracy:', accuracy)

# Predicting which industry will have high releases in 2022
predicted_releases = knn.predict(X_test22)
print("Petroleum refineries will have medium realeses in 2022 year based on its past ")